### PR TITLE
+Rotated and simpler OBC initialization

### PR DESCRIPTION
**Pull request by Hallberg-NOAA**
This PR consists of a series of 8 commits that simplify the calls used to initialize the OBCs, including initializing the OBCs directly on the rotated grid when grid rotation is in use.  This includes moving the code that initializes the OBCs out of `MOM_initialize_state()` and into the new subroutine `MOM_initialize_OBCs()`.  The code that sets up the OBCs was rearranged and consolidated into fewer blocks of code, including the combination of the functionality that had previously been provided by the separate OBC routines into other closely related routines and the elimination of some functionality that is no longer need with this simpler structure.  As a result, the publicly visible OBC interface to `rotate_OBC_init()`, `set_tracer_data()` and `open_boundary_setup_vert()` were eliminated altoghether. Several of the commits are intermediate steps introducing new subroutines that are subsequently removed, but these intermediate stages may be useful for debugging any problems that may arise.  There are also new calls to `open_boundary_end()` to manage OBC memory use.

  Most of the changes only relate to the case where grid rotation is being used for debugging,  and none of these changes will have any impact on cases that do not use open boundary conditions.  It has been verified that the OBCs still pass all of the grid rotation tests that they did previously.  All answers are bitwise identical, but there are several changes to publicly visible interfaces, including the addition of one new interface and the removal of three others.

The specific commits in this PR include:

 - NOAA-GFDL/MOM6@7bf96c63e +Open_boundary_setup_vert in init_segment_data
 - NOAA-GFDL/MOM6@c2a5382ee +Rotate_OBC_segment_fields in rotate_OBC_config
 - NOAA-GFDL/MOM6@e89d87495 Deallocate OBC memory with open_boundary_end
 - NOAA-GFDL/MOM6@b55748eed +Eliminate rotate_OBC_segment_tracer_registry
 - NOAA-GFDL/MOM6@44c6f2a0c +Rename and split rotate_OBC_init()
 - NOAA-GFDL/MOM6@62f478d1f Call MOM_initialize_OBCs with the rotated OBC type
 - NOAA-GFDL/MOM6@35e8cded3 Move calls out of MOM_initialize_OBCs
 - NOAA-GFDL/MOM6@39da6deb5 +Split MOM_initialize_OBCs from MOM_init_state

### CODE REVIEW

File: src/user/supercritical_initialization.F90 [hunk 1]
```diff
@@ -32,6 +33,8 @@ subroutine supercritical_set_OBC_data(OBC, G, GV, US, param_file)
   ! Local variables
   character(len=40)  :: mdl = "supercritical_set_OBC_data" ! This subroutine's name.
   real :: zonal_flow ! Inflow speed [L T-1 ~> m s-1]
+  integer :: unrot_dir ! The unrotated direction of the segmetn
```
Comment (kshedstrom) [inline src/user/supercritical_initialization.F90:36]: should be spelled segment

**Comment by Hallberg-NOAA** [comment]
Thank you for identifying and tracking down the segmentation faults in some cases with the OBC grid rotation, @kshedstrom.  I have put in PR #939 to address them, as well as another bug in the same section of code where we could have flipped the signs of the normal or tangential velocities twice, depending on which combinations of OBC fields are being read in on a face.

File: src/user/supercritical_initialization.F90 (hunk 1 above)
Comment (Hallberg-NOAA) [reply src/user/supercritical_initialization.F90]: Thank you for catching this spelling error.  It has been corrected.

**Review by marshallward — APPROVED** [review]

**Comment by Hallberg-NOAA** [comment]
Because this PR changes the order in the code with which `initialize_segment_data()` is called relative to other routines that call `get_param()`, the order with which some OBC-related parameters are logged in the MOM_parameter_doc files changes.  This PR has passed pipeline testing at https://gitlab.gfdl.noaa.gov/ogrp/mom6ci/MOM6/-/pipelines/28167 with the expected warnings about changes in order with which some parameters are logged in the MOM_parameter_doc files.
